### PR TITLE
Add support for GraalJS engine

### PR DESCRIPTION
**Pull request by SergiyIvan**
Added support for GraalJS engine.

### CODE REVIEW

File: engines/graaljs/extract.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+			cwd: tmpPath,
+		});
+		const installer = new Installer({
+			engine: binary,
+			path: tmpPath,
+		});
+		switch (os) {
+			case 'mac64':
+			case 'linux64': {
+				const executableName = `${fs.readdirSync(tmpPath).filter(file => file.startsWith('graaljs'))[0]}/bin/js`;
```
Comment (mathiasbynens) [inline engines/graaljs/extract.js:37]: Let’s use `.find()` instead of `instead of filter(fn)[0]`

File: engines/graaljs/extract.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+				installer.installBinary({ [executableName] : binary });
```
Comment (mathiasbynens) [inline engines/graaljs/extract.js:38]: style nit:

```suggestion
				installer.installBinary({ [executableName]: binary });
```

File: engines/graaljs/extract.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+				break;
+			}
+			case 'win64': {
+				const executableName = `${fs.readdirSync(tmpPath).filter(file => file.startsWith('graaljs'))[0]}\\bin\\js.exe`;
+				installer.installBinary(
+					{ [executableName] : `${binary}.exe` },
```
Comment (mathiasbynens) [inline engines/graaljs/extract.js:44]: style nit:

```suggestion
					{ [executableName]: `${binary}.exe` },
```

File: engines/graaljs/extract.js (hunk 1 above)
Comment (SergiyIvan) [reply engines/graaljs/extract.js]: Thanks for your comments, I've fixed that

**Comment by mathiasbynens** [comment]
Could you please sign the CLA?

**Comment by wirthi** [comment]
Hi @mathiasbynens 

@SergiyIvan contributes this PR as part of his work with Oracle. We are currently trying to figure out whether maybe Oracle has signed this already, or what our internal process would be to sign it. We cannot sign this individually.

Best,
Christian Wirth
Oracle Labs, GraalVM

**Comment by wirthi** [comment]
Hey @mathiasbynens anything else that @SergiyIvan should fix before this is ready to merge?

File: engines/graaljs/extract.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+const tar = require('tar');
+
+const { Installer } = require('../../shared/installer.js');
+const unzip = require('../../shared/unzip.js');
+
+const extract = ({ filePath, binary, alias, os }) => {
+	return new Promise(async (resolve, reject) => {
+		const tmpPath = path.dirname(filePath);
+		if(os.startsWith('win')) {
```
Comment (mathiasbynens) [inline engines/graaljs/extract.js:27]:
```suggestion
		if (os.startsWith('win')) {
```

to match the existing coding style

File: engines/graaljs/extract.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+			});
+		}
+		const installer = new Installer({
+			engine: binary,
+			path: tmpPath,
+		});
+		switch (os) {
+			case 'mac64':
+			case 'linux64': {
+				const executableName = `${fs.readdirSync(tmpPath).find(file => file.startsWith('graaljs'))}/bin/js`;
```
Comment (mathiasbynens) [inline engines/graaljs/extract.js:45]: There's a lot going on on this one line. Let's do `const foo = fs.read…;` and move that to its own line so it's easier to follow

File: engines/graaljs/extract.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+				installer.installBinary({ [executableName]: binary });
+				break;
+			}
+			case 'win64': {
+				const executableName = `${fs.readdirSync(tmpPath).find(file => file.startsWith('graaljs'))}\\bin\\js.exe`;
```
Comment (mathiasbynens) [inline engines/graaljs/extract.js:50]: same here

**Review by mathiasbynens — CHANGES_REQUESTED** [review]
Suggested a few minor changes. Once those are addressed, this looks good to go!

File: engines/graaljs/extract.js (hunk 4 above)
Comment (SergiyIvan) [reply engines/graaljs/extract.js]: Thanks, I've followed your suggestions. I've also rebased my branch and squashed all my commits into one

**Review by mathiasbynens — APPROVED** [review]

**Comment by mathiasbynens** [comment]
Thanks!

**Comment by mathiasbynens** [comment]
Published as part of `jsvu@1.12.0`.

**Pull request merged**